src/ansi-c/c_typecheck_code.cpp

```diff
@@ -842,35 +844,167 @@ void c_typecheck_baset::typecheck_dowhile(code_dowhilet &code)
+    error() << "assigns clause target must be an lvalue or " CPROVER_PREFIX
+               "POINTER_OBJECT expression"
+            << eom;
+  }
+
+  // Remark: an expression can be an lvalue and still contain side effects
+  // or ternary expressions. We detect them in a second step.
+  if(has_subexpr(target, ID_side_effect))
+  {
+    must_throw = true;
+    error().source_location = target.source_location();
+    error() << "side-effects not allowed in assigns clause targets" << eom;
+  }
+
+  if(has_subexpr(target, ID_if))
+  {
+    must_throw = true;
+    error().source_location = target.source_location();
+    error() << "ternary expressions not allowed in assigns "
+               "clause targets"
+            << eom;
+  }
+
+  if(must_throw)
+    throw 0;
+}
+
 void c_typecheck_baset::typecheck_spec_assigns(exprt::operandst &targets)
 {
+  exprt::operandst tmp;
+  bool must_throw = false;
+
   for(auto &target : targets)
   {
-    typecheck_expr(target);
-
-    if(target.type().id() == ID_empty)
+    if(!can_cast_expr<conditional_target_exprt>(target))
     {
+      must_throw = true;
       error().source_location = target.source_location();
-      error() << "void-typed targets not permitted in assigns clause" << eom;
-      throw 0;
-    }
-    else if(target.id() == ID_pointer_object)
-    {
-      // skip
+      error()
+        << "expected ID_conditional_target expression in assigns clause, found "
+        << id2string(target.id()) << eom;
     }
-    else if(!target.get_bool(ID_C_lvalue))
+
+    auto &conditional_target = to_conditional_target_expr(target);
+    validate_expr(conditional_target);
+
+    // typecheck condition
+    auto &condition = conditional_target.condition();
+    typecheck_spec_assigns_condition(condition);
+
+    if(condition.is_true())
     {
-      error().source_location = target.source_location();
-      error() << "non-lvalue target in assigns clause" << eom;
-      throw 0;
+      // if the condition is trivially true,
+      // simplify expr and expose the bare expressions
+      for(auto &actual_target : conditional_target.targets())
+      {
+        typecheck_spec_assigns_target(actual_target);
+        tmp.push_back(actual_target);
+      }
     }
-    else if(has_subexpr(target, ID_side_effect))
+    else
     {
-      error().source_location = target.source_location();
-      error() << "assigns clause is not side-effect free" << eom;
-      throw 0;
+      // if the condition is not trivially true,
+      // place each single target in its own conditional target expr,
+      // distributing the condition
+      for(auto &actual_target : conditional_target.targets())
+      {
+        typecheck_spec_assigns_target(actual_target);
+        exprt ops{ID_expression_list};
+        ops.add_to_operands(actual_target);
+        conditional_target_exprt result{condition, ops};
+        result.add_source_location() = conditional_target.source_location();
+        tmp.push_back(std::move(result));
```

Comment:
This appears to be the only place that constructs a `conditional_target_exprt`. I would, therefore, expect that `conditional_target_exprt` has a constructor that permit creating the desired expression with just a single statement, like this:
```
tmp.push_back(conditional_target_exprt{condition, exprt{ID_expression_list, typet{}, {actual_target}}, conditional_target.source_location()});
```